src/shell-window-tracker.c

```diff
@@ -176,6 +184,70 @@ gboolean shell_window_tracker_is_speedwagon_window (MetaWindow *window)
   return g_strcmp0 (meta_window_get_role (window), SPEEDWAGON_ROLE) == 0;
 }
 
+static gboolean
+watchdog_timeout_reached_cb (gpointer user_data)
+{
+  ShellWindowTracker *self = user_data;
+
+  self->watchdog_id = 0;
+  if (self->coding_app)
+    self->coding_app = NULL;
+  return FALSE;
+}
+
+/**
+ * shell_window_tracker_track_coding_app_window:
+ *
+ * Track a coding app.
+ *
+ */
+void shell_window_tracker_track_coding_app_window (ShellWindowTracker *tracker,
+                                                   MetaWindow *window)
+{
+  tracker->coding_app = window;
+  tracker->watchdog_id = g_timeout_add_seconds (WATCHDOG_TIMEOUT,
+                                                watchdog_timeout_reached_cb,
+                                                tracker);
+}
+
+/**
+ * shell_window_tracker_untrack_coding_app_window:
+ *
+ * Untrack a coding app.
+ *
+ */
+void shell_window_tracker_untrack_coding_app_window (ShellWindowTracker *tracker,
+                                                     MetaWindow *window)
+{
+  MetaWindow *coding_app;
+  coding_app = g_hash_table_lookup (tracker->builder_to_app, window);
+  if (coding_app) {
+    g_hash_table_remove (tracker->builder_to_app, window);
+  }
+}
+
+/**
+ * shell_window_tracker_get_app_from_builder:
+ * @tracker: An app monitor instance
+ * @builder: A #MetaWindow
+ *
+ * Returns: (transfer full): Application associated with Builder window
+ */
+
+MetaWindow *shell_window_tracker_get_app_from_builder (ShellWindowTracker *tracker,
+                                                       MetaWindow *builder)
+{
+  MetaWindow *app;
+
+  app = g_hash_table_lookup (tracker->builder_to_app, builder);
+  if (!app)
+    return NULL;
+
+  g_object_ref (app);
```

Comment:
Actually one of the functions in that file uses tranfer none. So I guess we can do that as well. If you prefer:

```
diff --git a/src/shell-window-tracker.c b/src/shell-window-tracker.c
index df8d2d1..73b5657 100644
--- a/src/shell-window-tracker.c
+++ b/src/shell-window-tracker.c
@@ -231,7 +231,7 @@ void shell_window_tracker_untrack_coding_app_window (ShellWindowTracker *tracker
  * @tracker: An app monitor instance
  * @builder: A #MetaWindow
  *
- * Returns: (transfer full): Application associated with Builder window
+ * Returns: (transfer none): Application associated with Builder window, or %NULL if none
  */
 
 MetaWindow *shell_window_tracker_get_app_from_builder (ShellWindowTracker *tracker,
@@ -243,8 +243,6 @@ MetaWindow *shell_window_tracker_get_app_from_builder (ShellWindowTracker *track
   if (!app)
     return NULL;
 
-  g_object_ref (app);
-
   return app;
 }
```